README.md
# BackSync

> Backup and sync your data with ease.

Supported backup:

- Local
- MongoDB
- MySQL

Supported sync:

- Local
- Google Drive
- Dropbox
- AWS S3

## Install

```bash
npm i -g @ayoubamine/backsync
```

## Usage

Generate config file `backsync.json`:

```bash
backsync init
```

Add backup source:

```bash
backsync add backup
```

Add sync source:

```bash
backsync add sync
```

Update default settings:

```bash
backsync settings
```

Run the backup and sync manually:

```bash
backsync run
```

Schedule the backup and sync in the [background](#process-management):

```bash
backsync schedule
```

Show the sync history:

```bash
backsync status
```

## Process Management

Install PM2 to run the process in the background:

```bash
npm i -g pm2
```

Create `ecosystem.config.js` file in the same directory that `backsync.json` exists and add the following content:

```js
module.exports = {
  apps: [
    {
      name: 'backsync',
      script: 'backsync',
      args: 'schedule',
    },
  ],
};
```

Then to start:

```bash
pm2 start ecosystem.config.js
```

Display the logs:

```bash
pm2 logs backsync
```

To stop and delete:

```bash
pm2 delete backsync
```

## Auth

### Google Drive

#### Step 1: Create service account

Follow these steps to create a service account and download your `credentials.json` file:
[Create service account](https://cloud.google.com/docs/authentication/production#create_service_account)

#### Step 2: Create backup folder

1. Go to <https://drive.google.com>.
2. Create new folder.
3. Right-click on the folder, select Share, add the service account email with the Editor permission and click Done.
4. Browse to the new folder and copy the folder id from the url. 'https://drive.google.com/drive/u/0/folders/FOLDER_ID'

### Dropbox

#### Step 1: Create an app in your Dropbox account

1. Go to <https://dropbox.com/developers/apps/create>.
2. Choose Scoped access on the first step.
3. Choose App folder on the second.
4. Give your app a name. That name will become a folder in your Dropbox account.
5. Click Create app.

#### Step 2: Generate access token

1. Go to Permissions tab and enable `files.content.write` scope.
1. Go to Settings tab, scroll down to OAuth 2 block, select 'No expiration' and click Generate.

### AWS S3

#### Step 1: AWS credentials

1. Go to <https://aws.amazon.com>.
2. Navigate to IAM > Users and click 'Add users'.
3. Fill the username, enable 'Access key - Programmatic access', click 'Next: Permissions', select 'Attach existing policies directly', enable the AmazonS3FullAccess permission and click 'Next: Tags' > 'Next: Review'.

#### Step 2: Create an S3 bucket

1. Go to <https://s3.console.aws.amazon.com>.
2. Click 'Create bucket', fill the bucket name and click 'Create bucket'.

## API

### Info

| Name        | Type     | Required |
| ----------- | -------- | -------- |
| name        | `string` | Yes      |
| description | `string` | No       |

### Backup

#### Local

##### Params

| Name | Type     | Required |
| ---- | -------- | -------- |
| path | `string` | Yes      |

##### Settings

| Name     | Type     | Required                    |
| -------- | -------- | --------------------------- |
| filename | `string` | No, Default: Original name. |

#### MongoDB

##### Params

| Name     | Type     | Required                                 |
| -------- | -------- | ---------------------------------------- |
| uri      | `string` | No, Default: `mongodb://localhost:27017` |
| database | `string` | Yes                                      |

##### Settings

| Name     | Type     | Required                    |
| -------- | -------- | --------------------------- |
| filename | `string` | No, Default: Database name. |

#### MySQL

##### Params

| Name     | Type     | Required                |
| -------- | -------- | ----------------------- |
| host     | `string` | No, Default: `hostname` |
| port     | `number` | No, Default: `3306`     |
| user     | `string` | No, Default: `root`     |
| password | `string` | No                      |
| database | `string` | Yes                     |

##### Settings

| Name     | Type     | Required                    |
| -------- | -------- | --------------------------- |
| filename | `string` | No, Default: Database name. |

### Sync

#### Local

| Name | Type     | Required |
| ---- | -------- | -------- |
| path | `string` | Yes      |

#### GDrive

| Name                         | Type     | Required |
| ---------------------------- | -------- | -------- |
| [credentials](#google-drive) | `string` | Yes      |
| [folderId](#google-drive)    | `string` | Yes      |

#### Dropbox

| Name              | Type     | Required |
| ----------------- | -------- | -------- |
| [token](#dropbox) | `string` | Yes      |

#### AWS S3

| Name                       | Type     | Required |
| -------------------------- | -------- | -------- |
| [accessKeyId](#aws-s3)     | `string` | Yes      |
| [secretAccessKey](#aws-s3) | `string` | Yes      |
| [bucketName](#aws-s3)      | `string` | Yes      |

### Settings

| Name     | Type     | Required                 |
| -------- | -------- | ------------------------ |
| cron     | `string` | No, Default: `0 0 * * *` |
| prefix   | `string` | No                       |
| filename | `string` | No                       |

## Contributions

Feel free to contribute to this project.

If you find a bug or want a feature, but don't know how to fix/implement it, please fill an [issue](https://github.com/ayoubamine/backsync/issues).
If you fixed a bug or implemented a new feature, please send a [pull request](https://github.com/ayoubamine/backsync/pulls).

## Changelog

[CHANGELOG](./CHANGELOG.md)

## License

[MIT License](./LICENSE.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.3.0]

- Customize the prefix and backup filename from settings.
- Fix the license link.

## [1.2.2]

- Roll back dependency inquirer to v8.1.3 <https://github.com/SBoudrias/Inquirer.js/issues/1046>.

## [1.2.1]

- Change the package name.
- Update dependencies.

## [1.2.0]

- Add AWS S3 sync source.

## [1.1.0]

- Add MySQL backup source.

## [1.0.0]

- Initial release.


lib/inquirer.js
const inquirer = require('inquirer');
const { PathPrompt } = require('inquirer-path');
const cleanDeep = require('clean-deep');
const chalk = require('chalk');
const cron = require('node-cron');
const path = require('path');

inquirer.registerPrompt('path', PathPrompt);

const { CRON_EXPRESSION } = require('./config');
const files = require('./utils/files');
const mongodb = require('./utils/mongodb');
const mysql = require('./utils/mysql');

module.exports = {
  askInfo() {
    return inquirer.prompt([
      {
        type: 'input',
        name: 'name',
        message: 'Enter your project name:',
        validate: (value) => !!value,
      },
      {
        type: 'input',
        name: 'description',
        message: 'Enter your project description:',
      },
    ]);
  },
  async askBackupSource() {
    const { type } = await inquirer.prompt([
      {
        type: 'list',
        name: 'type',
        choices: ['Local', 'MongoDB', 'MySQL'],
      },
    ]);

    let params;
    let settings;

    switch (type) {
      case 'Local': {
        params = await inquirer.prompt([
          {
            type: 'path',
            name: 'path',
            message: 'Enter the path of the file/directory:',
          },
        ]);

        const defaultValue = path.basename(params.path);
        const { filename } = await inquirer.prompt([
          {
            type: 'input',
            name: 'filename',
            message: 'Enter the filename of the backup file:',
            default: defaultValue,
          },
        ]);

        if (filename !== defaultValue) {
          settings = { filename };
        }
        break;
      }

      case 'MongoDB': {
        params = await inquirer.prompt([
          {
            type: 'input',
            name: 'uri',
            message: 'Enter the connection string:',
            default: 'mongodb://localhost:27017',
            validate: async (value) => {
              if (value.length) {
                const result = await mongodb.isValidConnectionString(value);

                return result;
              }

              return false;
            },
          },
        ]);

        const { database } = await inquirer.prompt([
          {
            type: 'list',
            name: 'database',
            message: 'Select the database name:',
            choices: await mongodb.getDatabaseNames(params.uri),
          },
        ]);

        params.database = database;

        const { filename } = await inquirer.prompt([
          {
            type: 'input',
            name: 'filename',
            message: 'Enter the filename of the backup file:',
            default: database,
          },
        ]);

        if (filename !== database) {
          settings = { filename };
        }
        break;
      }

      case 'MySQL': {
        params = await inquirer.prompt([
          {
            type: 'input',
            name: 'host',
            message: 'Enter the host:',
            default: 'localhost',
          },
          {
            type: 'number',
            name: 'port',
            message: 'Enter the port:',
            default: 3306,
          },
          {
            type: 'input',
            name: 'user',
            message: 'Enter the user:',
            default: 'root',
          },
          {
            type: 'input',
            name: 'password',
            message: 'Enter the password:',
          },
        ]);

        const isValid = await mysql.isValidCredentials(
          params.host,
          params.port,
          params.user,
          params.password
        );

        if (!isValid) {
          console.log(chalk.red('MySQL credentials is not valid!'));
          process.exit(1);
        }

        const { database } = await inquirer.prompt([
          {
            type: 'list',
            name: 'database',
            message: 'Select the database name:',
            choices: await mysql.getDatabaseNames(
              params.host,
              params.port,
              params.user,
              params.password
            ),
          },
        ]);

        params.database = database;

        const { filename } = await inquirer.prompt([
          {
            type: 'input',
            name: 'filename',
            message: 'Enter the filename of the backup file:',
            default: database,
          },
        ]);

        if (filename !== database) {
          settings = { filename };
        }
        break;
      }

      default:
        break;
    }

    return { type, params, settings };
  },
  async askSyncSource() {
    const { type } = await inquirer.prompt([
      {
        type: 'list',
        name: 'type',
        choices: ['Local', 'GDrive', 'Dropbox', 'AWS S3'],
      },
    ]);
    let params;

    switch (type) {
      case 'Local':
        params = await inquirer.prompt([
          {
            type: 'path',
            name: 'path',
            message: 'Enter the path of the backup directory:',
            directoryOnly: true,
          },
        ]);
        break;

      case 'GDrive':
        params = await inquirer.prompt([
          {
            type: 'path',
            name: 'credentials',
            message: 'Enter the path of the credentials file:',
            validate: (value) => {
              if (files.isFile(value)) {
                return true;
              }

              return 'Please enter a valid credentials file.';
            },
          },
          {
            type: 'input',
            name: 'folderId',
            message: 'Enter the folder id:',
            validate: (value) => !!value,
          },
        ]);
        break;

      case 'Dropbox':
        params = await inquirer.prompt([
          {
            type: 'input',
            name: 'token',
            message: 'Enter the token of your Dropbox account:',
            validate: (value) => !!value,
          },
        ]);
        break;

      case 'AWS S3':
        params = await inquirer.prompt([
          {
            type: 'input',
            name: 'accessKeyId',
            message: 'Enter the access key id:',
            validate: (value) => !!value,
          },
          {
            type: 'input',
            name: 'secretAccessKey',
            message: 'Enter the secret access key:',
            validate: (value) => !!value,
          },
          {
            type: 'input',
            name: 'bucketName',
            message: 'Enter the bucket name:',
            validate: (value) => !!value,
          },
        ]);
        break;

      default:
        break;
    }

    return { type, params };
  },
  async askSettings() {
    const settings = await inquirer.prompt([
      {
        type: 'input',
        name: 'cron',
        message: 'Enter your cron expression:',
        default: CRON_EXPRESSION,
        validate: (value) => cron.validate(value),
      },
      {
        type: 'input',
        name: 'prefix',
        message: 'Enter the prefix:',
      },
      {
        type: 'input',
        name: 'filename',
        message: 'Enter the backup filename:',
      },
    ]);

    return cleanDeep(settings);
  },
  async askSourceIndex(list) {
    const choices = list.map(({ type }, index) => `${index + 1} ${type}`);

    const { source } = await inquirer.prompt([
      {
        type: 'list',
        name: 'source',
        message: 'Select a source:',
        choices,
      },
    ]);
    const index = choices.indexOf(source);

    return { index };
  },
};
